server/athenian/api/auth.py

```diff
@@ -0,0 +1,119 @@
+import logging
+import os
+import re
+import json
+import aiohttp.web
+from urllib.request import urlopen
+from urllib.request import Request as HTTPRequest
+from jose import jwt
+
+# Format error response and append status code.
+class AuthError(Exception):
+    def __init__(self, error, status_code):
+        self.error = error
+        self.status_code = status_code
+
+class User():
+    def __init__(self, info):
+        self.username = info["nickname"]
+        self.fullname = info["name"]
+        self.picture = info["picture"]
+        self.updated = info["updated_at"]
+
+AUTH0_DOMAIN = os.environ.get("AUTH0_DOMAIN")
+AUTH0_AUDIENCE = os.environ.get("AUTH0_AUDIENCE")
+
+def AuthConfigured():
+    return AUTH0_DOMAIN and AUTH0_AUDIENCE
+
+class Auth0():
+    def __init__(self, domain=AUTH0_DOMAIN, audience=AUTH0_AUDIENCE, whitelist=tuple()):
+        self._domain = domain
+        self._audience = audience
+        self._algos = ["RS256"]
+        self._whitelist = whitelist
+
+    def _check_request(self, request):
+        for pattern in self._whitelist:
+            if re.match(pattern, request.path):
+                return True
+        return False
+
+    def _get_token_auth_header(self, request):
+        """Obtains the access token from the Authorization Header
+        """
+        auth = request.headers.get("Authorization", None)
+        if not auth:
+            raise AuthError("Authorization header is expected", 401)
+
+        parts = auth.split()
+
+        if parts[0].lower() != "bearer":
+            raise AuthError("Authorization header must start with Bearer", 401)
+        elif len(parts) == 1:
+            raise AuthError("Token not found", 401)
+        elif len(parts) > 2:
+            raise AuthError("Authorization header must be Bearer token", 401)
+
+        token = parts[1]
+        return token
+
+    def _get_user_info(self, token):
+        # TODO: cache based on decoded claims
+        jsonurl = urlopen(HTTPRequest("https://"+self._domain+"/userinfo", headers={"Authorization": "Bearer "+token}))
+        resp = json.loads(jsonurl.read())
+        return User(resp)
+
+    @aiohttp.web.middleware
+    async def middleware(self, request, handler):
+        if self._check_request(request):
+            return await handler(request)
+
+        # TODO: update periodically instead of pulling it on each request
+        jsonurl = urlopen("https://"+self._domain+"/.well-known/jwks.json")
+        jwks = json.loads(jsonurl.read())
+
+        try:
+            token = self._get_token_auth_header(request)
+            unverified_header = jwt.get_unverified_header(token)
+        except AuthError as e:
+            return aiohttp.web.Response(body=e.error, status=e.status_code)
+        except jwt.JWTError:
+            return aiohttp.web.Response(body="Invalid header. Use an RS256 signed JWT Access Token.", status=401)
+        if unverified_header["alg"] != self._algos[0]:
```

Comment:
```suggestion
        if unverified_header["alg"] != "RS256":
```
I mean, we already hardcode RS256 in the error message, so...